File: app/src/main/java/com/example/readative/room/dao/BookDao.java
package com.example.readative.room.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.example.readative.room.entity.Book;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;

@Dao
public interface BookDao {
    @Insert(onConflict = OnConflictStrategy.IGNORE)
    public Completable insertBooks(Book... books);

    @Update
    public Completable updateBooks(Book... books);

    @Delete
    public Completable deleteBooks(Book... books);

    @Query("SELECT * FROM book")
    public Flowable<List<Book>> loadAllBooks();

    @Query("SELECT * FROM book WHERE book_id = :id")
    public Flowable<Book> loadBookById(long id);

    @Query("SELECT book_id FROM book WHERE book_uuid = :uuid")
    public Single<Integer> loadBookIdByUUID(String uuid);

    @Query("SELECT * FROM book WHERE author_owner_id = :authorId")
    public Flowable<List<Book>> loadBooksByAuthor(long authorId);
}

File: data/local/src/main/java/com/example/local/dao/BookDao.kt
package com.example.local.dao

import androidx.room.*
import com.example.model.local.entity.Book
import com.example.model.local.result_entity.BookWithBasicInfo
import com.example.model.local.result_entity.BookWithFullInfo
import kotlinx.coroutines.flow.Flow

@Dao
interface BookDao {
    @Query("SELECT * FROM books")
    fun getBooks(): Flow<List<Book>>

    @Transaction
    @RewriteQueriesToDropUnusedColumns
    @Query("SELECT * FROM books")
    fun getBasicInfoBooks(): Flow<List<BookWithBasicInfo>>

    @Transaction
    @RewriteQueriesToDropUnusedColumns
    @Query("SELECT * FROM books")
    fun getFullInfoBooks(): Flow<List<BookWithFullInfo>>

    @Query("SELECT * FROM books WHERE id = :id")
    suspend fun getBookById(id: Long): Book?

    @Query("SELECT * FROM books WHERE checksum LIKE :checksum")
    suspend fun getBookByChecksum(checksum: String): Book?

    @Transaction
    @RewriteQueriesToDropUnusedColumns
    @Query("SELECT * FROM books WHERE checksum LIKE :checksum")
    suspend fun getBasicInfoBookByChecksum(checksum: String): BookWithBasicInfo?

    @Transaction
    @RewriteQueriesToDropUnusedColumns
    @Query("SELECT * FROM books WHERE id = :id")
    suspend fun getBasicInfoBookById(id: Long): BookWithBasicInfo?

    @Transaction
    @RewriteQueriesToDropUnusedColumns
    @Query("SELECT * FROM books WHERE id = :id")
    suspend fun getFullInfoBookById(id: Long): BookWithFullInfo?

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    suspend fun insertBook(book: Book): Long

    @Update
    suspend fun updateBook(book: Book)

    @Delete
    suspend fun deleteBook(book: Book)
}